### PR TITLE
Mask to spectrum

### PR DESCRIPTION
**Pull request by registerrier**
**Description**

This pull request adds `mask_safe` support in `MapDataset.to_spectrum_dataset`.

The `WcsNDMap.get_spectrum()` is modified to ensure dtype is conserved.

**Dear reviewer**

This should be ready. A test for `MapDatasetOnOff.to_spectrum_dataset()` might be added to make sure things are done consistently there as well.

### CODE REVIEW

File: gammapy/datasets/spectrum.py [hunk 1]
```diff
@@ -667,6 +667,7 @@ def info_dict(self, in_safe_energy_range=True):
         """
         info = dict()
         mask = self.mask_safe.data if in_safe_energy_range else slice(None)
+        print(self.mask_safe.data)
```
Comment (adonath) [inline gammapy/datasets/spectrum.py:670]: Remove print statement...

File: gammapy/datasets/tests/test_map.py [hunk 2]
```diff
@@ -201,6 +201,11 @@ def test_to_spectrum_dataset(sky_model, geom, geom_etrue, edisp_mode):
     spectrum_dataset_corrected = dataset_ref.to_spectrum_dataset(
         on_region, containment_correction=True
     )
+    mask = np.ones_like(dataset_ref.counts, dtype='bool')
+    mask[1,:,:]=0
```
Comment (adonath) [inline gammapy/datasets/tests/test_map.py:205]: Maybe define a spatial part for the mask as well?

File: gammapy/maps/wcsnd.py [hunk 3]
```diff
@@ -26,7 +26,7 @@
 
 
 class WcsNDMap(WcsMap):
-    """HEALPix map with any number of non-spatial dimensions.
+    """WCS map with any number of non-spatial dimensions.
```
Comment (adonath) [inline gammapy/maps/wcsnd.py:29]: Funny that this hasn't been noticed earlier...does it mean that users never look at docstrings? Or maybe they just don't care...:-)

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @registerrier, functionality wise this looks good to me!

However I think there is an efficiency concern for large images. E.g. in the `ExcessProfileEstimator` we call this methods O(1e2) time or more. Always creating a `mask_safe` if there is none and always multiplying the full array is maybe problematic. Instead I would propose the following solution, that we have e.g. in `.stack()` as well:
- Add a `weights` argument to `.get_spectrum()` and `.to_region_nd_map()`
- Check if weights is not None and take it into account on the cutout data if present. Either by making a second cutout for the weights or re-using the slices defined in `cutout.geom.cutout_info`
- The implementation in `.to_spectrum_dataset()` should then simplify to `.get_spectrum(..., weights=self.mask_safe)`

**Comment by registerrier** [comment]
A `weights` argument has been added to `WcsNDMap.get_spectrum` and `WcsNDMap.to_on_region`. It is assumed that the geometry is identical to the `Map` (contrary to `WcsNDMap.stack`.
The `dtype` issue in now dealt with using `np.any` instead of `np.sum` for the mask.

**Review by adonath — APPROVED** [review]
Thanks a lot @registerrier, no further comments from my side.

**Comment by adonath** [comment]
Fixes #2953.

**Pull request merged**